src/GafferScene/SceneAlgo.cpp

```diff
@@ -801,6 +804,175 @@ ScenePlug *SceneAlgo::sourceScene( GafferImage::ImagePlug *image )
+	parallelProcessLocations( scene, attributesFinder );
+	return result;
+}
+
+} // namespace
+
+IECore::PathMatcher GafferScene::SceneAlgo::linkedObjects( const ScenePlug *scene, const ScenePlug::ScenePath &light )
+{
+	PathMatcher lights;
+	lights.addPath( light );
+	return linkedObjects( scene, lights );
+}
+
+GAFFERSCENE_API IECore::PathMatcher GafferScene::SceneAlgo::linkedObjects( const ScenePlug *scene, const IECore::PathMatcher &lights )
+{
+	// We expect many locations to have the exact same expression for `linkedLights`,
+	// and evaluating the expression is fairly expensive. So we cache the results for
+	// sharing between locations. The cache only lives for the lifetime of this query.
+	using QueryCache = IECorePreview::LRUCache<std::string, bool, IECorePreview::LRUCachePolicy::TaskParallel>;
+	const Context *context = Context::current();
+	QueryCache queryCache(
+		[&lights, scene, context]( const std::string &setExpression, size_t &cost )
+		{
+			cost = 1;
+			Context::Scope scopedContext( context );
+			const IECore::PathMatcher linkedLights = SetAlgo::evaluateSetExpression( setExpression, scene );
+			for( PathMatcher::Iterator lightIt = lights.begin(), eIt = lights.end(); lightIt != eIt; ++lightIt )
+			{
+				if( linkedLights.match( *lightIt ) & PathMatcher::ExactMatch )
+				{
+					return true;
+				}
+			}
+			return false;
+		},
+		10000
+	);
+
+	IECore::PathMatcher result = findAttributes(
+		scene,
+		[&queryCache] ( const CompoundObject *fullAttributes ) {
+			auto *linkedLights = fullAttributes->member<StringData>( g_linkedLights );
+			return queryCache.get( linkedLights ? linkedLights->readable() : "defaultLights" );
+		}
+	);
+
+	result.removePaths( scene->set( g_lights )->readable() );
+	return result;
+}
+
+IECore::PathMatcher GafferScene::SceneAlgo::linkedLights( const ScenePlug *scene, const ScenePlug::ScenePath &object )
+{
+	IECore::ConstCompoundObjectPtr attributes = scene->fullAttributes( object );
+	auto *linkedLightsAttribute = attributes->member<StringData>( g_linkedLights );
+	const string linkedLights = linkedLightsAttribute ? linkedLightsAttribute->readable() : "defaultLights";
+	IECore::PathMatcher linkedPaths = SetAlgo::evaluateSetExpression( linkedLights, scene );
+	return linkedPaths.intersection( scene->set( g_lights )->readable() );
+}
+
+IECore::PathMatcher GafferScene::SceneAlgo::linkedLights( const ScenePlug *scene, const IECore::PathMatcher &objects )
+{
+	tbb::spin_mutex resultMutex;
+	IECore::PathMatcher result;
+	tbb::concurrent_unordered_set<std::string> processed;
+
+	auto functor = [&resultMutex, &result, &processed] ( const ScenePlug *scene, const ScenePlug::ScenePath &path ) {
+		IECore::ConstCompoundObjectPtr attributes = scene->fullAttributes( path );
+		auto *linkedLightsAttribute = attributes->member<StringData>( g_linkedLights );
+		const string linkedLights = linkedLightsAttribute ? linkedLightsAttribute->readable() : "defaultLights";
```

Comment:
I stared at this for a bit wondering why we wouldn't just call the above version of linkedLights for each ScenePath.  It makes sense that checking for duplicated strings on different objects is crucial for performance.  But this did lead to me contemplating ... is it better to do the intersection with the lights set before or after merging all the paths into the result?  I guess we're assuming that usually all the light paths will be lights, so we wouldn't save much ... oh and the same light can get hit from many different expressions, so it makes sense to only check it once at the end.  Just answered my own question, but maybe I'll leave it as evidence that I have actually read through this :P